### PR TITLE
[#19946] feat: add key pair QR code view

### LINKED ISSUE #19946: Allow users to view key-pair QR code from wallet settings
# Feature Issue

### User Story

As a user,
I want view the QR code of a key-pair while inside the wallet settings.

### Description 
Implement the action for displaying the QR code for a key-pair inside wallet settings.

### Designs

Figma action menu flow: [the Figma designs for the action menu](https://www.figma.com/file/hLFPHYnuM0pOzgiabtqQ6k/Descoped-Wallet-settings-for-Mobile?type=design&node-id=49-133839&mode=design&t=CtD5ivvGpzuXMOt0-4)

Figma QR code flow: [the Figma designs for the QR code](https://www.figma.com/file/hLFPHYnuM0pOzgiabtqQ6k/Descoped-Wallet-settings-for-Mobile?type=design&node-id=1919-19842&mode=design&t=CtD5ivvGpzuXMOt0-4)

### Notes
* Based on [the designs here](https://www.figma.com/file/hLFPHYnuM0pOzgiabtqQ6k/Descoped-Wallet-settings-for-Mobile?type=design&node-id=1919-19842&mode=design&t=CtD5ivvGpzuXMOt0-4), the user should be required to authenticate before revealing a key-pair's QR code. The authentication step should happen after the user confirmed to see the QR code by "sliding to reveal".

### Acceptance Criteria

| Screen | Description |
|--------|--------|
| Key-pair Actions Menu | Users can choose to show the encrypted QR code of a key-pair with the actions menu. |
| Key-pair Reveal QR Code | Users are asked to use the "slide to reveal" UI button to confirm that they want to view the QR code. |
| Key-pair Display QR Code | Users are able to see the QR code for the key-pair. |

### PR DESCRIPTION
**Pull request by mohsen-ghafouri**
fixes #19946

### Summary

Implement the action for displaying the QR code for a key-pair inside wallet settings.

#### Areas that maybe impacted

- Show Encrypted QR of key pair in wallet setting

### Steps to test

1. Have a paired device
2. Open Wallet setting
3. Click on Show encrypted QR of key pairs on device

### Result

https://github.com/status-im/status-mobile/assets/71308738/b8101f75-f9a4-4b84-9a0c-24a664a34e83

status: ready

### CODE REVIEW

File: src/status_im/contexts/settings/wallet/keypairs_and_accounts/actions/view.cljs [hunk 1]
```diff
@@ -7,13 +7,24 @@
...
-  [:<>
-   [quo/drawer-top props]
-   [quo/action-drawer
-    [(when (= (:type props) :keypair)
-       [{:icon                :i/edit
-         :accessibility-label :rename-key-pair
-         :label               (i18n/label :t/rename-key-pair)
-         :on-press            #(on-rename-request data)}])]]])
+  (let [has-paired-device (rf/sub [:pairing/has-paired-devices])]
+    (println has-paired-device)
```
Comment (smohamedjavid) [inline src/status_im/contexts/settings/wallet/keypairs_and_accounts/actions/view.cljs]: 💣

File: src/status_im/contexts/settings/wallet/keypairs_and_accounts/encrypted_qr/view.cljs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        :on-press   navigate-back}]
+      [rn/view {:style style/page-container}
+       [rn/view {:style style/title-container}
+        [quo/text
+         {:size   :heading-1
+          :weight :semi-bold
+          :style  {:color colors/white}}
+         (i18n/label :t/encrypted-key-pairs)]]
+       [rn/view {:style style/qr-container}
+        (if (sync-utils/valid-connection-string? code)
```
Comment (smohamedjavid) [inline src/status_im/contexts/settings/wallet/keypairs_and_accounts/encrypted_qr/view.cljs]: Suggestion: Can we move this check to a separate let var above it's repeated below in the code.

File: src/status_im/contexts/settings/wallet/events.cljs [hunk 3]
```diff
@@ -30,3 +34,26 @@
            :on-error   #(log/info "failed to rename keypair " %)}]]]})
 
 (rf/reg-event-fx :wallet/rename-keypair rename-keypair)
+
+(defn key-pair-export-connection-string
+  [{:keys [db]} [{:keys [sha3-pwd keypair-key-uid callback]}]]
+  (let [error             (get-in db [:profile/login :error])
+        handle-connection (fn [response]
+                            (println response)
```
Comment (smohamedjavid) [inline src/status_im/contexts/settings/wallet/events.cljs]: 💣

File: src/status_im/contexts/settings/wallet/events.cljs [hunk 4]
```diff
@@ -30,3 +34,26 @@
...
+                            (when (sync-utils/valid-connection-string? response)
+                              (callback response)
+                              (rf/dispatch [:hide-bottom-sheet])))]
+    (when-not (and error (string/blank? error))
+      (let [key-uid    (get-in db [:profile/profile :key-uid])
+            config-map (.stringify js/JSON
+                                   (clj->js {:senderConfig {:loggedInKeyUid key-uid
+                                                            :keystorePath (native-module/keystore-dir)
```
Comment (smohamedjavid) [inline src/status_im/contexts/settings/wallet/events.cljs]: Apologies 🙏 We need to add the key-uid of the logged-in user to the end for the sender.

We can get the keystore path and add the key-uid to the end, in the native module call in the `NetworkManager` file as we do for syncing and pass an empty string here.

Only for the receiver config, it's the root directory.

https://github.com/status-im/status-desktop/blob/d2f73cd88fa93046b5df2c7f49ba4d777fe3079d/src/app_service/service/devices/service.nim#L208

File: src/status_im/contexts/settings/wallet/events.cljs [hunk 5]
```diff
@@ -30,3 +34,26 @@
...
+
+(defn key-pair-export-connection-string
+  [{:keys [db]} [{:keys [sha3-pwd keypair-key-uid callback]}]]
+  (let [error             (get-in db [:profile/login :error])
+        handle-connection (fn [response]
+                            (println response)
+                            (when (sync-utils/valid-connection-string? response)
+                              (callback response)
+                              (rf/dispatch [:hide-bottom-sheet])))]
+    (when-not (and error (string/blank? error))
```
Comment (smohamedjavid) [inline src/status_im/contexts/settings/wallet/events.cljs]: I believe we don't have to do this check, as the standard authentication only calls the callback on successful verification of the password.

File: src/status_im/contexts/settings/wallet/events.cljs [hunk 6]
```diff
@@ -30,3 +34,26 @@
...
+      (let [key-uid    (get-in db [:profile/profile :key-uid])
+            config-map (.stringify js/JSON
+                                   (clj->js {:senderConfig {:loggedInKeyUid key-uid
```
Comment (smohamedjavid) [inline src/status_im/contexts/settings/wallet/events.cljs]: Suggestion: We can replace this chained transformation with the util function `clj->json` from `utils.transforms` ns which does the same.

File: src/status_im/contexts/settings/wallet/events.cljs [hunk 7]
```diff
@@ -30,3 +34,26 @@
...
+                                                            :keystorePath (native-module/keystore-dir)
+                                                            :keypairsToExport [keypair-key-uid]
+                                                            :password (security/safe-unmask-data
+                                                                       sha3-pwd)}
+                                             :serverConfig {:timeout 0}}))]
+        (native-module/get-connection-string-for-exporting-keypairs-keystores
+         config-map
+         handle-connection)))))
+
+(rf/reg-event-fx :wallet/get-key-pair-export-connection key-pair-export-connection-string)
```
Comment (smohamedjavid) [inline src/status_im/contexts/settings/wallet/events.cljs]: It's good we match the method name with the event name.

**Review by smohamedjavid — COMMENTED** [review]
Awesome work! 🙌

**Review by smohamedjavid — APPROVED** [review]
Nice work! 🙌

**Comment by mohsen-ghafouri** [comment]
Hey @status-im/mobile-qa could you please check test results?

**Review by seanstrom — APPROVED** [review]
Well done! ✅

**Comment by yevh-berdnyk** [comment]
hi @mohsen-ghafouri, thanks for your PR, e2e failures are known and not related. But quick manual testing would be nice to have here

**Comment by mohsen-ghafouri** [comment]
Hi @yevh-berdnyk this screen is still behind the feature flag, I think we agreed to have manual QA when we want to remove feature flag.

**Comment by pavloburykh** [comment]
> Hi @yevh-berdnyk this screen is still behind the feature flag, I think we agreed to have manual QA when we want to remove feature flag.

@mohsen-ghafouri thanks for clarification. In this case we can skip manual QA. PR is ready for merge.

**Pull request merged**